icu4c/source/common/locid.cpp

```diff
@@ -1340,7 +1340,25 @@ Locale::getUnicodeKeywordValue(StringPiece keywordName,
 void
 Locale::setKeywordValue(const char* keywordName, const char* keywordValue, UErrorCode &status)
 {
-    uloc_setKeywordValue(keywordName, keywordValue, fullName, ULOC_FULLNAME_CAPACITY, &status);
+    int32_t bufferLength = uprv_max((int32_t)(uprv_strlen(fullName) + 1), ULOC_FULLNAME_CAPACITY);
+    int32_t newLength = uloc_setKeywordValue(keywordName, keywordValue, fullName,
```

Comment:
Need to bail out early if the incoming status shows a failure. uloc_setKeywordValue(..., status) will return a meaningless value. And the incoming failure could be U_BUFFER_OVERFLOW_ERROR, which will further confuse the following logic.